polar-core/src/partial/partial.rs

```diff
@@ -84,16 +86,46 @@ impl Partial {
         self.add_constraint(op);
     }
 
-    pub fn isa(&mut self, other: Term) -> Box<dyn Runnable> {
-        let isa_op = op!(Isa, self.variable_term(), other);
-
-        let constraint_check = Box::new(IsaConstraintCheck::new(
-            self.constraints.clone(),
-            isa_op.clone(),
-        ));
+    pub fn isa(&mut self, other: Term) -> Option<Box<dyn Runnable>> {
+        match other.value() {
+            Value::Pattern(Pattern::Dictionary(fields)) => {
+                // Add field constraints.
+                for (field, value) in fields.fields.iter().rev() {
+                    self.add_constraint(op!(
+                        Unify,
+                        term!(op!(Dot, self.variable_term(), term!(field.clone()))),
+                        value.clone()
+                    ));
+                }
+                None
+            }
+            Value::Pattern(Pattern::Instance(InstanceLiteral { fields, tag })) => {
+                // Construct field-less Isa operation since field constraints will be added
+                // separately.
+                let tag_pattern = term!(pattern!(instance!(tag.clone())));
+                let type_constraint = op!(Isa, self.variable_term(), tag_pattern);
+                let existing = self.constraints.clone();
+                let check = Box::new(IsaConstraintCheck::new(existing, type_constraint.clone()));
+
+                self.add_constraint(type_constraint);
+
+                // Add field constraints.
```

Comment:
As an alternative, could we call `self.isa(Pattern::Dictionary)` here? Would save a small amount of code duplication. I can make this change & test, just lmk what you think.